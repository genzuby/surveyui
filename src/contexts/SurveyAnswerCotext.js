import React, { useState, createContext } from "react";

// Create context
export const SurveyContext = createContext(null);

export const SurveyAnswerStore = props => {
  const [answerInfo, setAnswerInfo] = useState({});

  return (
    <SurveyContext.Provider value={{ answerInfo, setAnswerInfo }}>
      {props.children}
    </SurveyContext.Provider>
  );
};
