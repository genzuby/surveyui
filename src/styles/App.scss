@import url("https://fonts.googleapis.com/css?family=Work+Sans&display=swap");

:root {
  --active--yellow: #f88f07;
  --light--yellow: #f7a02e;
  --available--bg: #44d350;
  --available--font: #fff;
  --available--border: #349c3d;
  --disable--bg: #e9e9e9;
  --disable--font: #9e9e9e;
  --disable--border: #c7c7c7;
  --text--main--color: #262626;
  --bg--dark--color: #241d4b;
  --bg--gray--color: #888888;
  --bg--gray--light: rgb(245, 245, 245);
  --bg--gray: rgb(235, 235, 235);
}

body,
html {
  font-family: "Work Sans", sans-serif;
  margin: 0;
  padding: 0;
  overflow: hidden;
  min-width: 320px;
}

.App {
  height: 100vh;
  position: relative;

  .SurveyBody {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    width: 94%;
    height: 96vh;
    border-radius: 1em;
    box-shadow: 2px 2px 2px rgba(0, 0, 0, 0.2);
    display: flex;
    align-items: center;
    flex-direction: column;

    .Header {
      height: 35%;
      .Heading {
        padding: 2em;
        h2 {
          text-align: center;
        }
      }
      .Content {
        padding: 2em 1.5em;

        p {
          text-align: center;
          font-size: 1.2rem;
        }
      }
    }

    .TypeComp {
      height: 40%;
      width: 100%;
      display: flex;
      align-items: flex-start;
      justify-content: center;
      flex-direction: column;
    }

    .Buttons {
      display: flex;
      align-items: center;
      justify-content: center;
      flex-direction: row;
      width: 95%;
      height: 25%;
    }
  }
}
