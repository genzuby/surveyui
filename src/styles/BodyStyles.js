/* eslint-disable */
import styled from "styled-components";

// SurveyBody BG MOBILE
export const BODY = styled.div`
  background-color: ${props =>
    props.first
      ? "rgba(0, 0, 0, 0.3)"
      : props.last
      ? "rgba(0, 0, 0, 0.55)"
      : ""};
  background: ${props =>
    props.first
      ? "url(/images/welcome.jpg)"
      : props.last
      ? "url(/images/thankyou.jpg)"
      : "var(--bg--gray--light)"};
  background-blend-mode: multiply;
  background-size: cover;
  background-repeat: no-repeat;
  color: ${props =>
    props.first ? "#fff" : props.last ? "#fff" : "var(--text--main--color)"};
  border: ${props => (props.first ? "0px" : props.last ? "0px" : "1px")} solid
    var(--bg--gray);
  p,
  h1 {
    text-shadow: ${props =>
      props.first
        ? "2px 2px 2px rgba(0,0,0,0.3)"
        : props.last
        ? "2px 2px 2px rgba(0,0,0,0.3)"
        : ""};
  }
`;

// SurveyBody BG WEB
export const BODYWEB = styled.div`
  background-color: ${props =>
    props.first
      ? "rgba(0, 0, 0, 0.3)"
      : props.last
      ? "rgba(0, 0, 0, 0.55)"
      : ""};
  background: ${props =>
    props.first
      ? "url(/images/welcome-web.jpg)"
      : props.last
      ? "url(/images/thankyou-web.jpg)"
      : "var(--bg--gray--light)"};
  background-blend-mode: multiply;
  background-size: cover;
  background-repeat: no-repeat;
  color: ${props =>
    props.first ? "#fff" : props.last ? "#fff" : "var(--text--main--color)"};
  border: ${props => (props.first ? "0px" : props.last ? "0px" : "1px")} solid
    var(--bg--gray);
  p,
  h1 {
    text-shadow: ${props =>
      props.first
        ? "2px 2px 2px rgba(0,0,0,0.3)"
        : props.last
        ? "2px 2px 2px rgba(0,0,0,0.3)"
        : ""};
  }
`;

// Previous / Nest Button
export const BUTTON = styled.button`
  width: 100%;
  background: ${props =>
    props.available ? "var(--available--bg)" : "var(--disable--bg)"};
  border: 1px solid
    ${props =>
      props.available ? "var(--available--border)" : "var(--disable--border)"};
  height: 2.6em;
  font-size: 1.1rem;
  color: ${props =>
    props.available ? "var(--available--font)" : "var(--disable--font)"};
  border-radius: 4px;
  margin: 0 0.2em;
  outline: none;
  /* cursor disable when it's not available */
  pointer-events: ${props => (!props.available ? "none" : "auto")};
  cursor: ${props => (!props.available ? "nnot-not-allowed" : "pointer")};
  text-align: center;
`;

// Input styles
export const INPUT = styled.input`
  width: 80%;
  height: 2.3em;
  font-size: 1.1rem;
  padding: 0 0.5em;
  border: 1px solid var(--bg--gray--color);
  position: absolute;
  left: 50%;
  transform: translate(-50%);
`;

export const INPUROTH = styled.div`
  padding: ${props => (props.type === "radio" ? ".8em 15%" : "1em")};
  label {
    font-size: 1rem;
  }
`;

export const INPUROTHWEB = styled.div`
  padding: ${props =>
    props.type === "radio"
      ? ".8em calc(25% + 5vw)"
      : "1em 0 1em calc(10% + 10vw)"};
  label {
    font-size: 1rem;
  }
`;
