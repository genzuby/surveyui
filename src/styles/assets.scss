@import "App.scss";

// Checkbox and radio input styling
input[type="checkbox"] {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  border: 0;
}
input[type="checkbox"] + label {
  display: inline-block;
  position: relative;
  cursor: pointer;
}
input[type="checkbox"] + label:before {
  content: " ";
  display: inline-block;
  width: 21px;
  height: 21px;
  line-height: 21px;
  margin: -2px 8px 0 0;
  text-align: center;
  vertical-align: middle;
  border: 1px solid var(--light--yellow);
  border-radius: 2px;
}

input[type="checkbox"] + label {
  padding-left: 30px;
}
input[type="checkbox"] + label:before {
  position: absolute;
  left: 0;
  top: 0;
  margin-top: 0;
  opacity: 0.6;
  box-shadow: none;
  border-color: var(--light--yellow);
  transition: all 0.12s, border-color 0.08s;
}

input[type="checkbox"]:checked + label:before {
  position: absolute;
  content: "";
  width: 10px;
  top: -5px;
  left: 5px;
  border-radius: 2px;
  opacity: 1;
  background: transparent;
  border-color: transparent var(--light--yellow) var(--light--yellow)
    transparent;
  border-top-color: transparent;
  border-left-color: transparent;
  transform: rotate(10deg);
  transition: transform 0.1s;
}

input[type="checkbox"]:checked + label:before {
  content: "\2714";
  top: 0;
  left: 0;
  width: 21px;
  line-height: 21px;
  color: var(--active--yellow);
  text-align: center;
  border: 0;
}

input[type="radio"] {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  border: 0;
}
input[type="radio"] + label {
  display: inline-block;
  position: relative;
  padding-left: 30px;
  cursor: pointer;
}
input[type="radio"] + label:before {
  content: "";
  position: absolute;
  left: 0;
  top: -4px;
  width: 21px;
  height: 21px;
  text-align: center;
  background: #fff;
  border: 1px solid var(--light--yellow);
  border-radius: 100%;
}

input[type="radio"]:checked + label:before {
  background: #fff;
  border-color: var(--light--yellow);
}
input[type="radio"]:checked + label:after {
  content: "";
  position: absolute;
  top: 1px;
  left: 5px;
  width: 13px;
  height: 13px;
  background: var(--active--yellow);
  border-radius: 100%;
}
