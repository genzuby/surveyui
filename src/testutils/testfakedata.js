export const surveylist = [
  {
    questionID: "page-welcome",
    type: "Page",
    heading: "Welcome to the survey",
    body: "You can start the survey by clicking the next button",
    next: "question-email"
  },
  {
    questionID: "page-uneligible",
    type: "Page",
    heading: "Not eligible",
    body: "We're sorry, but you are not eligible to use our producs/services"
  },
  {
    questionID: "page-finish",
    type: "Page",
    heading: "Thank you",
    body: "You have completed the survey"
  },
  {
    questionID: "question-email",
    type: "Text",
    question: "What is your email address?",
    next: "question-age"
  },
  {
    questionID: "question-age",
    type: "Radio",
    question: "Are you 18 or older?",
    choices: [
      {
        label: "Yes",
        value: "yes"
      },
      {
        label: "No",
        value: "no",
        next: "page-uneligible"
      }
    ],
    next: "question-problems"
  },
  {
    questionID: "question-problems",
    type: "Checkbox",
    question: "Do you suffer from any of the following problems?",
    choices: [
      {
        label: "I have dental implants",
        value: "dental_implants"
      },
      {
        label: "I have missing teeth",
        value: "missing_teeth"
      },
      {
        label: "I have unfilled cavities",
        value: "unfilled_cavities"
      },
      {
        label: "I have swollen or inflamed gums",
        value: "swollen_or_inflamed_gums"
      },
      {
        label: "I have retainers",
        value: "retainers"
      },
      {
        label: "None of the above",
        value: "no"
      }
    ],
    next: "page-finish"
  }
];

export const initData = [
  {
    surveyID: "345f52c7-78f6-4cc4-9f8a-0490744f7da3",
    surveyItems: [
      {
        questionID: "page-welcome",
        type: "Page",
        heading: "Welcome to the survey",
        body: "You can start the survey by clicking the next button",
        next: "question-email"
      },
      {
        questionID: "page-uneligible",
        type: "Page",
        heading: "Not eligible",
        body:
          "We're sorry, but you are not eligible to use our producs/services"
      },
      {
        questionID: "page-finish",
        type: "Page",
        heading: "Thank you",
        body: "You have completed the survey"
      },
      {
        questionID: "question-email",
        type: "Text",
        question: "What is your email address?",
        next: "question-age"
      },
      {
        questionID: "question-age",
        type: "Radio",
        question: "Are you 18 or older?",
        choices: [
          {
            label: "Yes",
            value: "yes"
          },
          {
            label: "No",
            value: "no",
            next: "page-uneligible"
          }
        ],
        next: "question-problems"
      },
      {
        questionID: "question-problems",
        type: "Checkbox",
        question: "Do you suffer from any of the following problems?",
        choices: [
          {
            label: "I have dental implants",
            value: "dental_implants"
          },
          {
            label: "I have missing teeth",
            value: "missing_teeth"
          },
          {
            label: "I have unfilled cavities",
            value: "unfilled_cavities"
          },
          {
            label: "I have swollen or inflamed gums",
            value: "swollen_or_inflamed_gums"
          },
          {
            label: "I have retainers",
            value: "retainers"
          },
          {
            label: "None of the above",
            value: "no"
          }
        ],
        next: "page-finish"
      }
    ]
  }
];

export const testAnswer = {
  userId: "test3333@gmail.com",
  userInfo: {
    under18: false,
    finished: true
  },
  answeredlist: {
    surveyID: "345f52c7-78f6-4cc4-9f8a-0490744f7da3",
    surveyItems: [
      {
        questionID: "question-problems",
        type: "Checkbox",
        choices: ["dental_implants", "unfilled_cavities"]
      }
    ]
  }
};
