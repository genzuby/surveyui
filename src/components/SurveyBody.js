import React, { useState, useContext, useEffect } from "react";
import { connect } from "react-redux";
// for Adapted web design
import { isMobile, isBrowser } from "react-device-detect";
// style component
import { BODY, BODYWEB } from "../styles/BodyStyles";
import { SurveyContext } from "../contexts/SurveyAnswerCotext";
import { answerSurvey } from "../actions";

// helper methods
import {
  getFirstPage,
  checkFirstLast,
  getPrevId,
  validateEmail
} from "./utils/utilFunctions";

// import components
import Buttons from "./Buttons";
import Content from "./Content";
import TypeComp from "./TypeComp";

const SurveyBody = ({ surveylist, answerSurvey }) => {
  const { answerInfo, setAnswerInfo } = useContext(SurveyContext);
  const [currentInfo, setCurrentInfo] = useState({});
  const [firstPage, setFirstFage] = useState(false);
  const [lastPage, setLastFage] = useState(false);
  const [inputValue, setInputValue] = useState([]);
  const [inputNext, setInputNext] = useState("");

  // set Inital context
  useEffect(() => {
    // when data fetched
    if (surveylist.length > 0) {
      const { surveyID, surveyItems } = surveylist[0];
      // check if there is context data, to check if inital survey or not
      if (!answerInfo.currentID) {
        // get first page id from fetched data
        const firstId = getFirstPage(surveyItems);
        // get detail data based on first question id
        const currentVal = surveyItems.find(
          item => item.questionID === firstId
        );
        // modify context info with basic info
        setAnswerInfo({ surveyID, currentID: firstId });
        //  first page or not
        setFirstFage(true);
        //  set status to distribute into to children
        setCurrentInfo(currentVal);
      }
    }
  }, [surveylist, answerInfo, currentInfo, setCurrentInfo, setAnswerInfo]);

  // on click previos button
  const onPrevButtonHandle = () => {
    if (surveylist.length > 0) {
      const { surveyItems } = surveylist[0];

      // Get Previous Information if it's not a first page
      if (!firstPage) {
        // Get previous id and info
        const prevID = getPrevId(surveyItems, currentInfo.questionID);
        const currentVal = surveyItems.find(item => item.questionID === prevID);

        // get Previous page Info and state update
        setCurrentInfo(currentVal);

        // find context info and setInputvalue
        if (prevID === "question-email") {
          setInputValue(answerInfo.email);
        } else if (prevID === "question-age") {
          setInputValue(answerInfo.over18);
        }

        //  save current Info into Context
        saveContextInfo("PREV", prevID);

        // check next page is the fitsr page ot not
        const position = checkFirstLast(surveyItems, prevID);
        position === "FIRST" ? setFirstFage(true) : setFirstFage(false);

        // not the last page anymore
        setLastFage(false);
      }
    }
  };

  // on click next button
  const onNextButtonHandle = () => {
    if (surveylist.length > 0) {
      const { surveyItems } = surveylist[0];
      if (!lastPage) {
        // Get next id
        const nextId = inputNext ? inputNext : currentInfo.next;

        // get next page Info and state update
        const currentVal = surveyItems.find(item => item.questionID === nextId);

        // validataion check
        if (!inputValidation()) {
          return;
        }

        setCurrentInfo({});
        setCurrentInfo(currentVal);

        //  save current Info into Context
        saveContextInfo("NEXT", nextId);

        // current position check
        const position = checkFirstLast(surveyItems, nextId);

        // when Last Page then update state
        if (position === "LAST") {
          setFirstFage(true);
          setLastFage(true);
        } else {
          setFirstFage(false);
          setLastFage(false);
        }

        // set inital state of input value and input next
        setInputValue([]);
        setInputNext("");

        //  if poistion is last postion insert data
        if (position === "LAST") {
          uploadDataToDB();
        }
      }
    }
  };

  // context info save
  const saveContextInfo = (option, targetId) => {
    setAnswerInfo({
      ...answerInfo,
      currentID: targetId
    });

    // when click next button and type is not page add data
    if (option === "NEXT" && currentInfo.type !== "Page") {
      if (currentInfo.questionID === "question-email") {
        setAnswerInfo(Object.assign(answerInfo, { email: inputValue }));
      } else if (currentInfo.questionID === "question-age") {
        setAnswerInfo(Object.assign(answerInfo, { over18: inputValue }));
      } else {
        setAnswerInfo(
          Object.assign(answerInfo, {
            questions: [
              {
                questionID: currentInfo.questionID,
                type: currentInfo.type,
                choices: inputValue
              }
            ]
          })
        );
      }
    }
  };

  // Data insert into DB
  const uploadDataToDB = () => {
    const answerBody = {
      userId: answerInfo.email,
      userInfo: {
        under18: inputNext ? true : false,
        finished: inputNext ? false : true
      },
      answeredlist: [
        {
          surveyID: answerInfo.surveyID,
          surveyItems: answerInfo.questions
        }
      ]
    };
    // post data
    answerSurvey(answerBody);
  };

  const inputValidation = () => {
    // find input value and cheack each item next value
    //  when found next value ==> display warning
    if (currentInfo.type === "Page") {
      return true;
    } else {
      if (inputValue.length === 0) {
        //  it here is input value(s)
        alert("Please input(or check) one(or more) item.");
        return false;
      } else {
        // email validation check
        if (currentInfo.questionID === "question-email") {
          if (validateEmail(inputValue)) {
            return true;
          } else {
            alert("Please input valid email address!");
            return false;
          }
        }
        return true;
      }
    }
  };

  const renderComponent = () => (
    <>
      <div className="Header" data-test="bodyComponentSection">
        {currentInfo.heading && (
          <h1 className="Heading">{currentInfo.heading}</h1>
        )}
        <Content
          contentVal={currentInfo.body || currentInfo.question}
          type={currentInfo.type}
        />
      </div>
      <TypeComp
        type={currentInfo.type}
        item={currentInfo}
        inputValue={inputValue}
        setInputValue={setInputValue}
        setInputNext={setInputNext}
        data-test="bodyComponentSection"
      />
      <Buttons
        first={firstPage}
        last={lastPage}
        onHandle={{ onPrevButtonHandle, onNextButtonHandle }}
        data-test="bodyComponentSection"
      />
    </>
  );

  return (
    <>
      {/* Mobile view */}
      {isMobile && (
        <BODY
          className="SurveyBody"
          // to make button disable, when the page get to the last, set first also true.
          // So need to back to normal.
          first={lastPage ? false : firstPage}
          last={lastPage}
        >
          {renderComponent()}
        </BODY>
      )}
      {/* web view */}
      {isBrowser && (
        <BODYWEB
          className="SurveyBody"
          first={lastPage ? false : firstPage}
          last={lastPage}
        >
          {renderComponent()}
        </BODYWEB>
      )}
    </>
  );
};

const mapStateToProps = ({ surveylist }) => {
  return { surveylist };
};

export default connect(
  mapStateToProps,
  { answerSurvey }
)(SurveyBody);
