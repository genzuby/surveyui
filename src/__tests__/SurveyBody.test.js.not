import React from "react";
import { shallow } from "enzyme";
import "jest-styled-components";
import "../testutils/setup";
import { findByTestAttr, testStore } from "../testutils";
import { initData, testAnswer } from "../testutils/testfakedata";
import SurveyBody from "../components/SurveyBody";
import { BODY, BODYWEB } from "../styles/BodyStyles";
import { SurveyContext } from "../contexts/SurveyAnswerCotext";

const setUp = (initialState = {}) => {
  const store = testStore(initialState);
  let getFirstPage, checkFirstLast, getPrevId, validateEmail;
  let answerInfo, setAnswerInfo;
  const props = {
    getFirstPage
  };
  const wrapper = shallow(
    <SurveyContext.Provider value={{ answerInfo, setAnswerInfo }}>
      <SurveyBody store={store} {...props} />
    </SurveyContext.Provider>
  ).dive();
  return wrapper;
};

describe("Survey Component", () => {
  let wrapper;

  beforeEach(() => {
    const initialState = { surveylist: initData, answer: testAnswer };
    wrapper = setUp(initialState);
  });

  describe("Should Render without errors", () => {
    it("renders correctly", () => {
      // const instance = wrapper.instance();

      console.log(wrapper.debug())

      expect(wrapper.find('.SurveyBody')).toHaveLength(1);
    });

    it("Should render 3 div as children", () => {
      const sections = findByTestAttr(wrapper, "bodyComponentSection");
      expect(sections).toHaveLength(3);
    });
  });

  describe("Should do correct actions", () => {
    it("Should display first page info correctly in a Mobile device", () => {
      wrapper.setProps({
        getFirstPage: "page-welcome"
      });

      expect(wrapper.find(BODY)).toHaveStyleRule(
        "background",
        "url(/images/welcome.jpg)"
      );
    });

    it("Should display first page info correctly in a Browser device", () => {
      wrapper.setProps({
        getFirstPage: "page-welcome"
      });

      expect(wrapper.find(BODYWEB)).toHaveStyleRule(
        "background",
        "url(/images/welcome-web.jpg)"
      );
    });

    it("Should display last page info correctly in a Mobile device", () => {
      wrapper.setProps({
        getFirstPage: "page-welcome"
      });

      expect(wrapper.find(BODY)).toHaveStyleRule(
        "background",
        "url(/images/thankyou.jpg)"
      );
    });

    it("Should display last page info correctly in a Browser device", () => {
      wrapper.setProps({
        getFirstPage: "page-welcome"
      });

      expect(wrapper.find(BODYWEB)).toHaveStyleRule(
        "background",
        "url(/images/thankyou-web.jpg)"
      );
    });
  });
});
