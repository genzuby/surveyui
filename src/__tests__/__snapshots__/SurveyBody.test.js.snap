// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Survey Component Should Render without errors renders correctly 1`] = `undefined`;
